OpenRA.Mods.Common/Activities/Air/FlyAttack.cs

```diff
@@ -118,8 +132,33 @@ public override Activity Tick(Actor self)
 					return NextActivity;
 				}
 
-				// Fly towards the last known position
-				QueueChild(self, new Fly(self, target, WDist.Zero, lastVisibleMaximumRange, checkTarget.CenterPosition, Color.Red), true);
+				// Fly towards the desired facing (if CanHover) or last known position
+				if (aircraft.Info.CanHover)
+					aircraft.SetPosition(self, aircraft.CenterPosition + aircraft.FlyStep(desiredFacing));
+				else
+					QueueChild(self, new Fly(self, target, WDist.Zero, lastVisibleMaximumRange, checkTarget.CenterPosition, Color.Red), true);
+
+				return this;
+			}
+
+			if (aircraft.Info.CanHover)
+			{
+				// Fly towards the target
+				if (!target.IsInRange(pos, attackAircraft.GetMaximumRangeVersusTarget(target)))
+					aircraft.SetPosition(self, aircraft.CenterPosition + aircraft.FlyStep(desiredFacing));
+
+				// Fly backwards from the target
+				if (target.IsInRange(pos, attackAircraft.GetMinimumRangeVersusTarget(target)))
+				{
+					// Facing 0 doesn't work with the following position change
+					var facing = 1;
+					if (desiredFacing != 0)
+						facing = desiredFacing;
+					else if (aircraft.Facing != 0)
+						facing = aircraft.Facing;
+					aircraft.SetPosition(self, aircraft.CenterPosition + aircraft.FlyStep(-facing));
+				}
```

Comment:
Maybe this complete branch should queue `HeliFly` as a `ChildActivity` instead of doing all the positioning stuff itself?